cpp_linter/common_fs/file_filter.py

```diff
@@ -0,0 +1,216 @@
+import configparser
+import os
+from pathlib import Path, PurePath
+from typing import List, Optional, Set
+from . import FileObj
+from ..loggers import logger
+
+
+class FileFilter:
+    """A reusable mechanism for parsing and validating file filters.
+
+    :param extensions: A list of file extensions in which to focus.
+    :param ignore_value: The user input specified via :std:option:`--ignore`
+        CLI argument.
+    :param not_ignored: A list of files or paths that will be explicitly not ignored.
+    :param tool_specific_name: A clang tool name for which the file filter is
+        specifically applied. This only gets used in debug statements.
+    """
+
+    def __init__(
+        self,
+        ignore_value: str = "",
+        extensions: Optional[List[str]] = None,
+        not_ignored: Optional[List[str]] = None,
+        tool_specific_name: Optional[str] = None,
+    ) -> None:
+        #: A set of file extensions that are considered C/C++ sources.
+        self.extensions = set(extensions or [])
+        #: A set of ignore patterns.
+        self.ignored: Set[str] = set()
+        #: A set of not-ignore patterns.
+        self.not_ignored: Set[str] = set(not_ignored or [])
+        self._tool_name = tool_specific_name or ""
+        self._parse_ignore_option(paths=ignore_value)
+
+    def parse_submodules(self, path: str = ".gitmodules"):
+        """Automatically detect submodules from given ``path``.
+        This will add each submodule to the `ignored` list unless already specified as
+        `not_ignored`."""
+        git_modules = Path(path)
+        if git_modules.exists():
+            submodules = configparser.ConfigParser()
+            submodules.read(git_modules.resolve().as_posix())
+            for module in submodules.sections():
+                sub_mod_path = submodules[module]["path"]
+                if not self.is_file_in_list(
+                    ignored=False, file_name=PurePath(sub_mod_path)
+                ):
+                    logger.info(
+                        "Appending submodule to ignored paths: %s", sub_mod_path
+                    )
+                    self.ignored.add(sub_mod_path)
```

Comment:
The submodule ignored path should be relative to the .gitmodules file's location. Currently this function only works well at repo-root.